### PR TITLE
Close #279: Add Http4s interop

### LINKED ISSUE #279: Add http4s server support
Few months ago we were drop blaze support. Currently only one backed for use is akka. It's not fair. Let's add [http4s](https://http4s.org) support.

### PR DESCRIPTION
**Pull request by morozov11**
added http4s server support

### CODE REVIEW

File: interop/http4s/src/main/scala/korolev/http4s/package.scala [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  (implicit ec: ExecutionContext): HttpRoutes[F] = {
+
+    val PATH_PREFIX = "bridge/web-socket"
+
+    val dsl: Http4sDsl[F] = Http4sDsl[F]
+    import dsl._
+
+    val r = HttpRoutes.of[F] {
+
+      case req @ GET -> path  =>
```
Comment (fomkin) [inline interop/http4s/src/main/scala/korolev/http4s/package.scala:49]: Why GET only?

File: interop/http4s/src/main/scala/korolev/http4s/FS2StreamSubscriber.scala [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+import EffectRunner._
+import scala.collection.immutable.Queue
+
+/** Implementation of a `org.reactivestreams.Subscriber`.
+ *
+ * This is used to obtain a `fs2.Stream` from an upstream reactivestreams system.
+ *
+ * @see [[https://github.com/reactive-streams/reactive-streams-jvm#2-subscriber-code]]
+ */
+final class FS2StreamSubscriber[F[_]: ConcurrentEffect, A](val sub: FS2StreamSubscriber.FSM[F, A])
```
Comment (fomkin) [inline interop/http4s/src/main/scala/korolev/http4s/FS2StreamSubscriber.scala:18]: I suppose we do not need this anymore.

File: build.sbt [hunk 3]
```diff
@@ -147,6 +147,26 @@ lazy val akka = project
...
+    normalizedName := "korolev-http4s",
+    libraryDependencies ++= Seq(
+      "org.http4s"     %% "http4s-blaze-server" % "0.21.14",
+      "org.http4s"     %% "http4s-circe"        % "0.21.14",
+      "org.http4s"     %% "http4s-dsl"          % "0.21.14",
+      "org.http4s"     %% "http4s-blaze-client" % "0.21.14",
+      "org.typelevel"  %% "cats-effect" % "2.3.1",
+      "co.fs2"         %% "fs2-core" % "2.5.0",
+      "co.fs2"         %% "fs2-io" % "2.5.0",
+      "co.fs2"         %% "fs2-reactive-streams" % "2.5.0"
```
Comment (fomkin) [inline build.sbt:165]: Do need this anymore

File: build.sbt [hunk 4]
```diff
@@ -147,6 +147,26 @@ lazy val akka = project
...
+  .settings(
+    normalizedName := "korolev-http4s",
+    libraryDependencies ++= Seq(
+      "org.http4s"     %% "http4s-blaze-server" % "0.21.14",
+      "org.http4s"     %% "http4s-circe"        % "0.21.14",
+      "org.http4s"     %% "http4s-dsl"          % "0.21.14",
+      "org.http4s"     %% "http4s-blaze-client" % "0.21.14",
+      "org.typelevel"  %% "cats-effect" % "2.3.1",
+      "co.fs2"         %% "fs2-core" % "2.5.0",
+      "co.fs2"         %% "fs2-io" % "2.5.0",
```
Comment (fomkin) [inline build.sbt:164]: What you are use from fs2-io?

File: build.sbt [hunk 5]
```diff
@@ -147,6 +147,26 @@ lazy val akka = project
...
+  .enablePlugins(GitVersioning)
+  .settings(crossVersionSettings)
+  .settings(commonSettings: _*)
+  .settings(
+    normalizedName := "korolev-http4s",
+    libraryDependencies ++= Seq(
+      "org.http4s"     %% "http4s-blaze-server" % "0.21.14",
+      "org.http4s"     %% "http4s-circe"        % "0.21.14",
+      "org.http4s"     %% "http4s-dsl"          % "0.21.14",
+      "org.http4s"     %% "http4s-blaze-client" % "0.21.14",
```
Comment (fomkin) [inline build.sbt:161]: Do not need this.

File: build.sbt [hunk 6]
```diff
@@ -147,6 +147,26 @@ lazy val akka = project
...
 
+lazy val http4s = project
+  .in(interop / "http4s")
+  .enablePlugins(GitVersioning)
+  .settings(crossVersionSettings)
+  .settings(commonSettings: _*)
+  .settings(
+    normalizedName := "korolev-http4s",
+    libraryDependencies ++= Seq(
+      "org.http4s"     %% "http4s-blaze-server" % "0.21.14",
```
Comment (fomkin) [inline build.sbt:158]: http4s supports few backends as I know.

File: interop/http4s/src/main/scala/korolev/http4s/package.scala [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        println("ws get")
```
Comment (fomkin) [inline interop/http4s/src/main/scala/korolev/http4s/package.scala:47]: remove this

File: build.sbt (hunk 6 above)
Comment (morozov11) [reply build.sbt:158]: only http4s-blaze-server of them supports websocket

File: interop/http4s/src/main/scala/korolev/http4s/package.scala [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    val dsl: Http4sDsl[F] = Http4sDsl[F]
+    import dsl._
+    HttpRoutes.of[F] {
+
+      case req @ GET -> path =>
+        val body = LazyBytes.empty[F]
+        val korolevRequest = mkKorolevRequest(req, path.toString, body)
+        handleHttpResponse(korolevServer, korolevRequest)
+
+      case req @ POST -> path =>
```
Comment (fomkin) [inline interop/http4s/src/main/scala/korolev/http4s/package.scala:100]: `case req =>`

File: interop/http4s/src/main/scala/korolev/http4s/package.scala [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  private def configureWsRoute[F[_]: Effect: ConcurrentEffect, S: StateSerializer: StateDeserializer, M]
+  (korolevServer: KorolevService[F])
+  (implicit ec: ExecutionContext): HttpRoutes[F] = {
+
+    val dsl: Http4sDsl[F] = Http4sDsl[F]
+    import dsl._
+
+    HttpRoutes.of[F] {
+
+      case req @ GET -> path  =>
```
Comment (fomkin) [inline interop/http4s/src/main/scala/korolev/http4s/package.scala:47]: As I understand any GET request interpreted as WS. I do not see where you check ws handshake request.

**Comment by fomkin** [comment]
Need documentation like https://fomkin.org/korolev/user-guide.html#_akka

File: interop/http4s/src/main/scala/korolev/http4s/package.scala (hunk 9 above)
Comment (morozov11) [reply interop/http4s/src/main/scala/korolev/http4s/package.scala:47]: The path to the websocket is defined by  'val WS_PATH_PREFIX = "bridge/web-socket"'. All other GET requests routed to http handler. I don't know how to do it better.

File: docs/user-guide.adoc [hunk 10]
```diff
@@ -906,6 +906,48 @@ This integration overrides default logging behavior to `akka.event.Logging`.
...
+------------------------------
+
+[source,scala]
+------------------------------
+val service = KorolevServiceConfig[...](
+  ...
+))
+------------------------------
+
+And create Korolev route, parametrized with ZIO effect type:
```
Comment (fomkin) [inline docs/user-guide.adoc:925]: I think we should point that http4s integration can work with every effect processor which supports cats-effect.

File: interop/http4s/src/main/scala/korolev/http4s/package.scala (hunk 9 above)
Comment (fomkin) [reply interop/http4s/src/main/scala/korolev/http4s/package.scala:48]: 👍

**Pull request merged**